examples/async2-passthru.rs

```diff
@@ -0,0 +1,447 @@
+struct Win32EventFuture {
+    #[allow(dead_code)]
+    notif: Win32EventNotification, // The Win32 event notification object.
+    waker: Arc<AtomicWaker>,       // An atomic waker for waking the future.
+    ready: Arc<AtomicBool>,        // An atomic boolean indicating whether the event is ready.
+}
+
+impl Win32EventFuture {
+    /// Create a new `Win32EventFuture` instance with the specified event handle.
+    fn new(event_handle: HANDLE) -> Result<Self> {
+        let waker = Arc::new(AtomicWaker::new());
+        let ready = Arc::new(AtomicBool::new(false));
+
+        Ok(Self {
+            waker: waker.clone(),
+            ready: ready.clone(),
+            notif: Win32EventNotification::new(
+                event_handle,
+                Box::new(move |_| {
+                    println!("=========================================================================================================================================");
+                    ready.store(true, Ordering::SeqCst);
+                    waker.wake();
+                    unsafe { ResetEvent(event_handle) };
+                }),
+            )?,
+        })
+    }
+
+    /// Poll for the packet event.
+    pub fn poll_packet_event(&mut self, cx: &mut Context) -> Poll<Result<()>> {
+        if self.ready.swap(false, Ordering::Relaxed) {
+            Poll::Ready(Ok(()))
+        } else {
+            self.waker.register(cx.waker());
+            Poll::Pending
+        }
+    }
+}
+
+impl Future for Win32EventFuture {
+    type Output = Result<()>;
+
+    /// Poll the future to check if the event is ready.
+    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        Pin::into_inner(self).poll_packet_event(cx)
+    }
+}
+/// Win32 event notifications
+struct Win32EventNotification {
+    win32_event: HANDLE,               // The Win32 event handle.
+    wait_object: HANDLE,               // The wait object handle.
+    callback: *mut Win32EventCallback, // A pointer to the Win32 event callback function.
+}
+
+impl std::fmt::Debug for Win32EventNotification {
+    /// Implementing the Debug trait for the Win32EventNotification struct.
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Win32EventNotification: {:?}", self.wait_object)
+    }
+}
+
+type Win32EventCallback = Box<dyn Fn(BOOLEAN) + Send>; // A type alias for the Win32 event callback function.
+
+impl Win32EventNotification {
+    /// Register for Win32 event notifications.
+    fn new(win32_event: HANDLE, cb: Win32EventCallback) -> Result<Self> {
+        // Defining the global callback function for the Win32 event.
+        unsafe extern "system" fn global_callback(caller_context: *mut c_void, time_out: BOOLEAN) {
+            (**(caller_context as *mut Win32EventCallback))(time_out)
+        }
+
+        let callback = Box::into_raw(Box::new(cb)); // Creating a raw pointer to the callback function.
```

Comment:
I never said it was ideal :)

It's been a while since I touched that code so I didn't know that it is also in there. Can we get rid of the double allocation?

Reply:
I don't know for sure, if simply remove it then results types mismatch.

Reply:
In summary, I should keep the code as it is, with two levels of boxing, because we are dealing with raw pointers. The outer box is turned into a raw pointer, which is then cast back into a box in the callback function. The inner box is used to store the closure on the heap.